Release k6 v0.35.0 (#2227)
grafana/k6: release notes/v0.35.0.md

k6 v0.35.0 is here! :tada: It introduces several new features that nicely enhance its usability and also contains a whole lot of fixes and ongoing efforts with refactoring.

> In total, we have closed 14 issues. We have also branched out three new xk6 extensions during this release :star:

Reviewer: ```suggestion
In total, we have closed 14 issues. We have also branched out three new [xk6 extensions](https://k6.io/docs/misc/k6-extensions/) during this release :star:
```

Doesn't hurt to always link to the docs, some users might have never heard about xk6 :man_shrugging:

Author: It makes sense but now that I've added it, this actually looks strange to me as I'd think the link leads to those three extensions :sweat_smile:

## New features

### Ability to set VU-wide custom metric tags ([#2172](https://github.com/grafana/k6/pull/2172))

k6 now supports setting tags for VUs as part of the Execution API with an easy key-value interface. These tags are attached to the metrics emitted by the VU. Example usage:

```js
import http from 'k6/http';
import exec from 'k6/execution';

export const options = {
    duration: '10s',
    vus: 3,
};

export default function () {
    exec.vu.tags['mytag'] = 'value';
    exec.vu.tags['vuId'] = exec.vu.idInTest;

    console.log(`mytag is ${exec.vu.tags['mytag']} and my VU's ID in tags ${exec.vu.tags['vuId']}`);

>     // the metrics these HTTP requests will get tagged with `mytag` and `vuId`:

Reviewer: ```suggestion
    // the metrics these HTTP requests emit will get tagged with `mytag` and `vuId`:
```

    http.batch(['https://test.k6.io', 'https://test-api.k6.io']);
}
```

One of the most requested use cases for this feature is that now we can tag all metrics with the current stage number. [With a bit of JS code](https://github.com/grafana/k6/issues/796#issuecomment-959396841) it is possible to calculate which stage of a [`ramping-vus`](https://k6.io/docs/using-k6/scenarios/executors/ramping-vus/) or [`ramping-arrival-rate`](https://k6.io/docs/using-k6/scenarios/executors/ramping-arrival-rate/) scenario the VU is currently in. This in turn allows the setting of [thresholds](https://k6.io/docs/using-k6/thresholds/) only on the metrics that were emitted in specific stages of the test run! :tada:

> There are some caveats, however: values can be only of `String`, `Number` or `Boolean` type, while values of other types will result either in a warning or an exception. Additionally, given that k6 has a whole bunch of system tags, one should be careful with using them as keys. You can read complete information about VU tags in [`k6/execution docs`](https://k6.io/docs/javascript-api/k6-execution/#vu).

Reviewer: ```suggestion
There are some caveats, however: values can be only of `String`, `Number` or `Boolean` type, while values of other types will result either in a warning or an exception (if [`throw`](https://k6.io/docs/using-k6/options/#throw) is enabled). Additionally, given that k6 has a whole bunch of system tags, one should be careful with using them as keys. You can read complete information about VU tags in [`k6/execution docs`](https://k6.io/docs/javascript-api/k6-execution/#vu).
```

### Initial basic support for JS promises

> With the goja update in [#2197](https://github.com/grafana/k6/pull/2197), you can now make a `Promise` and chaining in your k6 scripts:

Reviewer: ```suggestion
With the goja update in [#2197](https://github.com/grafana/k6/pull/2197), you can now make a `Promise` and chain it in your k6 scripts:
```

```js
export default function () {
    var p = new Promise((resolve, reject) => {
        console.log('do something promising!');
        reject('here');
    });

    p.then(
        (s) => { console.log('fulfilled with', s) },
        (s) => { console.log('rejected with', s) },
    );
}
```

It must be noted that `Promise`s are not used by k6 itself yet but it is a stepping stone for implementing async functionality in future releases. Thanks, @dop251, for your awesome work in developing [goja](https://github.com/dop251/goja)! :heart:

### Support for gRPC server reflection ([#2160](https://github.com/grafana/k6/pull/2160))

> k6's GRPC was extended with a support for server reflection which allows one to use gRPC even without `proto` file at hand. In other words, the following script is now possible:

Reviewer: ```suggestion
k6's gRPC capabilities were extended with support for server reflection, which allows one to use gRPC even without a `proto` file at hand. In other words, the following script is now possible:
```

```js
import grpc from 'k6/net/grpc';
import { check } from "k6";

let client = new grpc.Client();

export default () => {

>     client.connect("127.0.0.1:10000", { plaintext: true })

Reviewer: ```suggestion
    client.connect("127.0.0.1:10000", { plaintext: true, reflect: true })
```

        reject('here');
    });

    p.then(
        (s) => { console.log('fulfilled with', s) },
        (s) => { console.log('rejected with', s) },
    );
}
```

It must be noted that `Promise`s are not used by k6 itself yet but this addition is a stepping stone for implementing async functionality in future releases. Thanks, @dop251, for your awesome work in developing [goja](https://github.com/dop251/goja)! :heart:

### Support for gRPC server reflection ([#2160](https://github.com/grafana/k6/pull/2160))

k6's gRPC capabilities were extended with a support for server reflection which allows one to use gRPC even without a `proto` file at hand. In other words, the following script is now possible:

```js
import grpc from 'k6/net/grpc';
import { check } from "k6";

let client = new grpc.Client();

export default () => {
	client.connect("127.0.0.1:10000", {plaintext: true, reflect: true})
	const response = client.invoke("main.RouteGuide/GetFeature", {
		latitude: 410248224,
		longitude: -747127767
	})

	check(response, {"status is OK": (r) => r && r.status === grpc.StatusOK});
	console.log(JSON.stringify(response.message))

	client.close()
}
```

> You can read more about the protocol [here](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) Thanks, @joshcarp, for putting a lot of effort into this feature!

Reviewer: ```suggestion
You can read more about the protocol [here](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md). Thanks, @joshcarp, for putting a lot of effort into this feature!
```

## Breaking changes

- Addition of common metrics registry no longer allows defining custom metrics with the same name as one of the builtin metrics, e.g. `new Counter("http_req_duration")` will abort. Builtin metrics may no longer be referenced as global objects in xk6 extensions either ([#1832](https://github.com/grafana/k6/issues/1832)).
- Fix inconsistency in environment variables' names: use `K6_NO_SETUP` and `K6_NO_TEARDOWN` options instead of `NO_SETUP` and `NO_TEARDOWN` ([#2140](https://github.com/grafana/k6/pull/2140))

## Bugs fixed!

- Fix of a misleading sorting of custom submetrics ([#2198](https://github.com/grafana/k6/pull/2198)). Thanks, @knittl!
- Fix for `afero.FS` bug which caused some `afero`-dependent xk6 extensions to fail ([#2216](https://github.com/grafana/k6/pull/2216)).
- Fix for websockets bug when `User-Agent` setting was ignored ([#2151](https://github.com/grafana/k6/pull/2151)). Thanks, @cooliscool!
- Fixes for tests, Github actions, and Loki integration ([#2205](https://github.com/grafana/k6/pull/2205), [#2153](https://github.com/grafana/k6/pull/2153), [#2220](https://github.com/grafana/k6/pull/2220)).

## Maintenance

> - Update a whole bunch of dependencies ([#2159](https://github.com/grafana/k6/pull/2159), [#2170](https://github.com/grafana/k6/pull/2170), [#2165](https://github.com/grafana/k6/pull/2165)).

Author: Those are golang lib updates and alpine image update: they're useful but it seems to me they shouldn't impact average user much. The 3rd issue here is another goja update -> template literals. I didn't re-check by hand but it was mentioned somewhere in chats that feature was supported even before, just with Babel, so it means it's no major change from user's perspective.

> - Update a whole bunch of dependencies ([#2159](https://github.com/grafana/k6/pull/2159), [#2170](https://github.com/grafana/k6/pull/2170), [#2165](https://github.com/grafana/k6/pull/2165)).

Reviewer 2: > that feature was supported even before

probably is bit better supported as usual as Babel has it's limits and is a bit old now. I don't think we should mention it for that, but mostly as this means that babel will now transpile even less stuff which means that it's a bit faster. For the octal/binary literals that meant for a specific script that was taking over 3 minutes to transpile that it went down to just under 1minute, which is quite considerable change for the people who have *big* scripts with specific features, in this case I think a big POST body with binary/octal literals.

The rest of the changes are probably mostly fixes for http2 which got a ton of fixes in this release

k6 v0.35.0 is here! :tada: It introduces several new features that nicely enhance usage of k6 as well as contains a whole number of fixes and ongoing efforts with refactoring.

> In total, we have closed 14 issues and merged 43 PRs. We have also branched out three new xk6 extensions during this release :star:

Author: Yes, those numbers are the ones labeled with the milestone.
If it's the norm not to label PRs with milestone based on their usefulness (is it?) then yes, not sure if stats would makes sense at all :confused: They definitely won't be sustainable then. I can remove it now, but it's something to think about in the future probably.

## New features

### VU tags ([#2172](https://github.com/grafana/k6/pull/2172))

> k6 now supports setting tags for VUs as part of the Execution API with an easy key-value interface. Example usage:

Reviewer 2: I feel like this doesn't explain that metrics emitted by this VU will be tagged with those tags, and I am not certain if that is as obvious to everybody as it is to us

Author: Small clarification won't hurt probably :+1:

```js
import exec from 'k6/execution';
import { sleep } from 'k6';

export const options = {
    duration: '10s',
    vus: 3,
};

export default function () {
    exec.vu.tags['mytag'] = 'value';
    exec.vu.tags['vuId'] = exec.vu.idInTest;

>     console.log("mytag is", exec.vu.tags['mytag'], 'and my VU\'s ID in tags', exec.vu.tags['vuId']);

Reviewer 3: ```suggestion
    console.log(`mytag is ${exec.vu.tags['mytag']}, and my VU's ID in tags ${exec.vu.tags['vuId']}`;
```
Sorry for spotting it only now, but I think a template literal could improve it.

Author: Yes, I agree; ~ "get used to JS fancy stuff" :smile:

    sleep(5);
}
```

> There are some caveats, however; using `Object` or `Array` as a value will result in a warning or an exception if `--throw` option is used. Additionally, given that k6 has a whole bunch of system tags, one should be careful with using them as keys. You can read complete information about VU tags in [`k6/execution docs`](https://k6.io/docs/javascript-api/k6-execution/#vu).

Author: Hm, I wanted to avoid repeating the docs too much here but still make a warning... But listing only supported types is probably a good middle-ground. So it'll be smth like this:
```
There are some caveats, however: values can be only of `String`, `Number` or `Boolean` type, while values of other types will result either in a warning or an exception. Additionally, given that k6 has a whole bunch of system tags, one should be careful with using them as keys. You can read complete information about VU tags in [`k6/execution docs`](https://k6.io/docs/javascript-api/k6-execution/#vu).
```

### JS promises are now an option 

With the goja update in [#2197](https://github.com/grafana/k6/pull/2197), you can now make a `Promise` and chaining in your k6 scripts:

> 

Author: "Useless" is a bit too much: then it should be removed completely :slightly_smiling_face:. I'll add something about "stepping stone for async", etc.

export default function () {
    var p = new Promise((resolve, reject) => {
        console.log('do something promising!');
        reject('here');
    });

    p.then(
        (s) => { console.log('fulfilled with', s) },
        (s) => { console.log('rejected with', s) },
    );
}
```

### InfluxDB v2 API

After hard work at working out how to integrate InfluxDB v2 API, it was decided to pull that integration into an [xk6-output-influxdb extension](https://github.com/grafana/xk6-output-influxdb). InfluxDB v1 API is still available as before, with minor optimizations ([#2190](https://github.com/grafana/k6/pull/2190)).

### New xk6 extensions

With this release, we have added two brand-new xk6 extensions:

- [xk6-browser](https://github.com/grafana/xk6-browser) is a browser automation extension which relies on [Chrome Devtools Protocol](https://chromedevtools.github.io/devtools-protocol/).
- [xk6-output-remote-write](https://github.com/grafana/xk6-output-prometheus-remote) is an output extension with support for [Prometheus Remote-Write protocol](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit#heading=h.3p42p5s8n0ui).

### Other notable changes

> - k6's GRPC was extended with a support for server reflection which allows one to use GRPC even without `proto` file at hand. You can read more about the protocol [here](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) ([#2160](https://github.com/grafana/k6/pull/2160)).

Reviewer 2: we in general thank the people (generally authors) that helped with a feature at the end of the line after the PR number for [example](https://github.com/grafana/k6/blob/v0.34.0/release%20notes/v0.33.0.md?plain=1#L34)

Reviewer 2: we also do that even if they did not make the final PR, they still did help and in this particular case AFAIK they did 99% of the work

Author: I wasn't the reviewer for that PR and didn't realize it was 99% :confused: This suggests that this way of writing release notes may not be the best TBH.

> - Support for cookies in `k6/ws` ([#2193](https://github.com/grafana/k6/pull/2193)).

Reviewer 2: ```suggestion
- Support for cookie jars in `k6/ws` ([#2193](https://github.com/grafana/k6/pull/2193)).
```

> - Forbid `metric.Add` to let `NaN` values through ([#1876](https://github.com/grafana/k6/pull/1876)); instead, print out nice warnings or throw an exception if `--throw` is enabled. These warnings will not become too large ([#2219](https://github.com/grafana/k6/pull/2219)).

Reviewer 2: Not certain it's needed to point out those warnings will not become too large, as this is not a bugfix for something we have released and I would argue most people would not have it as problem :)

Reviewer: Yeah, we can skip `These warnings will not become too large`, we can just add the second PR in the brackets, like this:
```suggestion
- Forbid `metric.Add` to let `NaN` values for custom metrics through. Instead, k6 will log nice warnings or throw an exception if `--throw` is enabled ([#1876](https://github.com/grafana/k6/pull/1876), [#2219](https://github.com/grafana/k6/pull/2219)).
```

- Support for compression in websockets ([#2162](https://github.com/grafana/k6/pull/2162)). Thanks, @cooliscool!

## Breaking changes

> - Addition of common metrics registry no longer allows defining custom metrics with the same name as one of the builtin metrics, e.g. `new Counter("http_req_duration")` will abort. Builtin metrics may no longer be referenced as global objects in xk6 extensions either ([#1832](https://github.com/grafana/k6/issues/1832)).

Reviewer 2: it isn't about just built-in metrics, it's about all metrics. If you try to redefine a metric with the same name but with different types it will error out. While the non built-in case is likely a bug it was also previously not watched. 

also is there a reason why here specifically you link to the issues not the pull request?

Author: No, that's just a mistake :slightly_smiling_face: Thanks!

- Fix inconsistency in environment variables' names: use `K6_NO_SETUP` and `K6_NO_TEARDOWN` options instead of `NO_SETUP` and `NO_TEARDOWN` ([#2140](https://github.com/grafana/k6/pull/2140)).

## Bugs fixed!

- Fix of a misleading sorting of custom submetrics ([#2198](https://github.com/grafana/k6/pull/2198)). Thanks, @knittl!

> - Fix for `afero.FS` bug which caused some `afero`-dependent xk6 extensions to fail ([#2216](https://github.com/grafana/k6/pull/2216)).

Reviewer 2: ```suggestion
- Implement a newer version of the `afero.FS` interface for internal filesystems, so that extension depending on that or newer version can be built ([#2216](https://github.com/grafana/k6/pull/2216)).
```

> - Fix for websockets bug when `User-Agent` setting was ignored ([#2151](https://github.com/grafana/k6/pull/2151)). Thanks, @cooliscool!

Reviewer 2: ```suggestion
- websockets now uses the global `User-Agent` setting ([#2151](https://github.com/grafana/k6/pull/2151)). Thanks, @cooliscool!
```
seems more natural to me :shrug:

k6 v0.35.0 is here! :tada: It introduces several new features that nicely enhance usage of k6 as well as contains a whole number of fixes and ongoing efforts with refactoring.

In total, we have closed 14 issues. We have also branched out three new xk6 extensions during this release :star:

## New features

> ### VU tags ([#2172](https://github.com/grafana/k6/pull/2172))

Reviewer: ```suggestion
### Ability to set VU-wide custom metric tags ([#2172](https://github.com/grafana/k6/pull/2172))
```

### JS promises are now an option 

With the goja update in [#2197](https://github.com/grafana/k6/pull/2197), you can now make a `Promise` and chaining in your k6 scripts:

```js
export default function () {
    var p = new Promise((resolve, reject) => {
        console.log('do something promising!');
        reject('here');
    });

    p.then(
        (s) => { console.log('fulfilled with', s) },
        (s) => { console.log('rejected with', s) },
    );
}
```

It must be noted that `Promise`s are not used by k6 itself yet but it is a stepping stone for implementing async functionality in the future releases.

### InfluxDB v2 API

After hard work at working out how to integrate InfluxDB v2 API, it was decided to pull that integration into an [xk6-output-influxdb extension](https://github.com/grafana/xk6-output-influxdb). InfluxDB v1 API is still available as before, with minor optimizations ([#2190](https://github.com/grafana/k6/pull/2190)).

> ### New xk6 extensions

Reviewer: :thinking: this should probably be its own major section, and it should probably be below `Other notable changes`. Also, the actual extension names should be more prominent, headers in their own right instead of a simple list. And `xk6-output-influxdb` should probably be here, not in the changes above.

Something like this:
```
## New features
### Ability to set VU-wide custom metric tags
### Initial basic support for JS promises
### Support gRPC server reflection
### Other notable changes and improvements

## New xk6 extensions
### xk6-browser
### xk6-output-remote-write
### xk6-output-influxdb
```

### Other notable changes

- k6's GRPC was extended with a support for server reflection which allows one to use GRPC even without `proto` file at hand. You can read more about the protocol [here](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) ([#2160](https://github.com/grafana/k6/pull/2160)). Thanks, @joshcarp!
- Support for cookie jars in `k6/ws` ([#2193](https://github.com/grafana/k6/pull/2193)).
- Forbid `metric.Add` to let `NaN` values through ([#1876](https://github.com/grafana/k6/pull/1876)); instead, print out nice warnings or throw an exception if `--throw` is enabled. These warnings will not become too large ([#2219](https://github.com/grafana/k6/pull/2219)).
- Support for compression in websockets ([#2162](https://github.com/grafana/k6/pull/2162)). Thanks, @cooliscool!
- Switch to camel case for CLI options to the outputs ([#2150](https://github.com/grafana/k6/pull/2150)). Thanks, @JosephWoodward!

## Breaking changes

> - Addition of common metrics registry no longer allows defining custom metrics with the same name as one of the builtin metrics, e.g. `new Counter("http_req_duration")` will abort. Similarly, an attempt to redefine a metric with the same name but with different type will error out. Builtin metrics may no longer be referenced as global objects in xk6 extensions either ([#2071](https://github.com/grafana/k6/pull/2071)).

Reviewer: ```suggestion
- The addition of common metrics registry ([#2071](https://github.com/grafana/k6/pull/2071)) no longer allows defining custom metrics with the same name as one of the builtin metrics, e.g. `new Counter("http_req_duration")` will now abort. Similarly, an attempt to redefine a metric with the same name but with different type will error out. Builtin metrics may no longer be referenced as global objects in xk6 extensions either.
```

- Fix inconsistency in environment variables' names: use `K6_NO_SETUP` and `K6_NO_TEARDOWN` options instead of `NO_SETUP` and `NO_TEARDOWN` ([#2140](https://github.com/grafana/k6/pull/2140)).

> 

Reviewer 2: We should mention https://github.com/grafana/k6/pull/2234, probably with some very small not and that it only matters for extension developers. AFAIK nobody outside of us is using the interfaces that were changed at this time

## Bugs fixed!

> - Fix of a misleading sorting of custom submetrics ([#2198](https://github.com/grafana/k6/pull/2198)). Thanks, @knittl!

Reviewer: ```suggestion
- Fix of a misleading sorting of custom submetrics in the default end-of-test summary ([#2198](https://github.com/grafana/k6/pull/2198)). Thanks, @knittl!
```

- Fix for extensions depending on `afero.FS`: implement a newer version of the `afero.FS` interface for internal filesystems so that extension depending on that or newer version can be built ([#2216](https://github.com/grafana/k6/pull/2216)).
- Fix for websockets: websockets now use the global `User-Agent` setting ([#2151](https://github.com/grafana/k6/pull/2151)). Thanks, @cooliscool!
- Fixes for tests, Github actions, and Loki integration ([#2205](https://github.com/grafana/k6/pull/2205), [#2153](https://github.com/grafana/k6/pull/2153), [#2220](https://github.com/grafana/k6/pull/2220)).

## Maintenance

- Update a whole bunch of dependencies ([#2159](https://github.com/grafana/k6/pull/2159), [#2170](https://github.com/grafana/k6/pull/2170), [#2165](https://github.com/grafana/k6/pull/2165)).
- Switch to Go 1.17 ([#2156](https://github.com/grafana/k6/pull/2156)). Thanks, @b5710546232!
- Get rid of `mapstructure` ([#2223](https://github.com/grafana/k6/pull/2223)).

> - Much neater error message on `nil` response body ([#2195](https://github.com/grafana/k6/pull/2195)). Thanks, @daniel-shuy!

Reviewer: This should probably be moved to the `Other notable changes and improvements` section, it has user-facing changes :man_shrugging:

Author: Not sure about "notable" here but if GRPC is important enough to be moved out into a separate part, that section with "Other notable changes" no longer seems as large :sweat_smile:

Reviewer: hmm yeah... this section used to be called `Enhancements and UX improvements` in old release notes, so maybe `Other changes and UX improvements`? :man_shrugging:

Author: :+1: yes, that seems more fitting now

Author: Update is up.
